src/rust/catnap/queue.rs

```diff
@@ -6,85 +6,220 @@
+    pub fn new(qtype: QType, socket: Socket) -> Self {
         Self {
             qtype,
-            fd,
-            socket: Socket::new(),
-            pending_ops: HashMap::<TaskHandle, YielderHandle>::new(),
+            socket: Rc::new(RefCell::new(socket)),
+            pending_ops: Rc::new(RefCell::new(HashMap::<TaskHandle, YielderHandle>::new())),
+        }
+    }
+
+    pub fn local(&self) -> Option<SocketAddrV4> {
+        self.socket.borrow().local()
+    }
+
+    pub fn remote(&self) -> Option<SocketAddrV4> {
+        self.socket.borrow().remote()
+    }
+
+    pub fn bind(&mut self, local: SocketAddrV4) -> Result<(), Fail> {
+        self.socket.borrow_mut().bind(local)
+    }
+
+    pub fn listen(&mut self, backlog: usize) -> Result<(), Fail> {
+        self.socket.borrow_mut().listen(backlog)
+    }
+
+    pub fn set_accept(&mut self) -> Result<(), Fail> {
+        self.socket.borrow_mut().accept()
+    }
+
+    pub async fn accept(&mut self, yielder: Yielder) -> Result<Self, Fail> {
+        loop {
+            let mut socket = self.socket.borrow_mut();
+            match socket.try_accept() {
+                Ok(socket) => {
+                    break Ok(Self {
+                        qtype: self.qtype,
+                        socket: Rc::new(RefCell::new(socket)),
+                        pending_ops: Rc::new(RefCell::new(HashMap::<TaskHandle, YielderHandle>::new())),
+                    })
+                },
+                Err(Fail { errno, cause: _ }) if errno == libc::EWOULDBLOCK || errno == libc::EAGAIN => {
+                    // Operation in progress. Check if cancelled.
+                    drop(socket);
+                    if let Err(e) = yielder.yield_once().await {
+                        break Err(e);
+                    }
+                },
+                Err(e) => break Err(e),
+            }
         }
     }
 
-    /// Get underlying POSIX file descriptor.
-    pub fn get_fd(&self) -> Option<RawFd> {
-        self.fd
+    pub fn set_connect(&mut self) -> Result<(), Fail> {
+        self.socket.borrow_mut().connect()
     }
 
-    /// Set underlying POSIX file descriptor.
-    pub fn set_fd(&mut self, fd: RawFd) {
-        self.fd = Some(fd);
+    pub async fn connect(&mut self, remote: SocketAddrV4, yielder: Yielder) -> Result<(), Fail> {
+        loop {
+            let mut socket = self.socket.borrow_mut();
+            match socket.try_connect(remote) {
+                Ok(r) => break Ok(r),
+                Err(Fail { errno, cause: _ }) if retry_errno(errno) => {
+                    // Operation in progress. Check if cancelled.
+                    drop(socket);
+                    if let Err(e) = yielder.yield_once().await {
+                        break Err(e);
+                    }
+                },
+                Err(e) => break Err(e),
+            }
+        }
+    }
+
+    pub fn close(&mut self) -> Result<(), Fail> {
+        self.socket.borrow_mut().close()?;
```

Comment:
What would be the implications of borrowing the socket here just once for both calls?

Reply:
I was just being lazy, have changed to a single borrow.